### PR TITLE
James 3401 Invalid mailbox filters don't return error in new JMAP

### PR DESCRIPTION
**Pull request by nvduc91**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 1]
```diff
@@ -5397,6 +5397,45 @@ trait EmailQueryMethodContract {
...
+        .contentType(JSON)
+        .extract
+        .body
+        .asString
+
+      assertThatJson(response)
+        .inPath("methodResponses[0][1]")
+        .isEqualTo(
+        """{
+         |     "type":"invalidArguments","description":"{\"errors\":[{\"path\":\"obj.filter\",\"messages\":[\"Unsupported filter\"]}]}"}
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:5435]: Can we have linebreaks to make this JSON more readable?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailQuerySerializer.scala [hunk 2]
```diff
@@ -81,7 +81,11 @@ class EmailQuerySerializer @Inject()(mailboxIdFactory: MailboxId.Factory) {
     case _ => JsError(s"Expecting a JsString to represent a known operator")
   }
 
-  private implicit val filterConditionReads: Reads[FilterCondition] = Json.reads[FilterCondition]
+  private implicit val filterConditionReads: Reads[FilterCondition] = {
+    case JsObject(underlying) if underlying.keySet.diff(FilterCondition.SUPPORTED).nonEmpty => JsError(s"Unsupported filter")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailQuerySerializer.scala:85]: Can we say which filters were unsupported?

`JsError(s"Unsupported filter ${underlying.keySet.diff(FilterCondition.SUPPORTED).toSting}")`

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/MailboxQuerySerializer.scala [hunk 3]
```diff
@@ -40,7 +40,12 @@ object MailboxQuerySerializer {
       .getOrElse(JsError(s"$value is not a valid role"))
     case _ => JsError("Expecting a JsString to be representing a role")
   }
-  private implicit val filterReads: Reads[MailboxFilter] = Json.reads[MailboxFilter]
+
+  private implicit val filterReads: Reads[MailboxFilter] =  {
+    case JsObject(underlying) if underlying.keySet.diff(MailboxFilter.SUPPORTED).nonEmpty => JsError(s"Unsupported filter")
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/MailboxQuerySerializer.scala:45]: Can we say which filters were unsupported?

`JsError(s"Unsupported filter ${underlying.keySet.diff(MailboxFilter.SUPPORTED).toSting}")`

**Review by chibenwa — COMMENTED** [review]
Looks good to me!

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxQueryMethodContract.scala [hunk 4]
```diff
@@ -437,6 +437,45 @@ trait MailboxQueryMethodContract {
...
+        .contentType(JSON)
+        .extract
+        .body
+        .asString
+
+      assertThatJson(response)
+        .inPath("methodResponses[0][1]")
+        .isEqualTo(
+        """{
+         |     "type":"invalidArguments","description":"{\"errors\":[{\"path\":\"obj.filter\",\"messages\":[\"Unsupported filter\"]}]}"}
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxQueryMethodContract.scala:475]: Line breaks here too?

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
```
[b0fa3d2463f014de25008593c17def63761f47e1] [ERROR] Tests run: 10, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.409 s <<< FAILURE! - in org.apache.james.jmap.rfc8621.memory.MemoryMailboxQueryMethodTest
[b0fa3d2463f014de25008593c17def63761f47e1] [ERROR] shouldReturnInvalidArgumentsWhenInvalidFilterCondition  Time elapsed: 0.248 s  <<< ERROR!
[b0fa3d2463f014de25008593c17def63761f47e1] java.lang.IllegalArgumentException: Can not parse expected value.
[b0fa3d2463f014de25008593c17def63761f47e1] Caused by: com.fasterxml.jackson.core.JsonParseException: 
[b0fa3d2463f014de25008593c17def63761f47e1] Unexpected character ('p' (code 112)): was expecting comma to separate Object entries
[b0fa3d2463f014de25008593c17def63761f47e1]  at [Source: (StringReader); line: 3, column: 35]
```
related I think :)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailQuerySerializer.scala [hunk 5]
```diff
@@ -81,7 +81,18 @@ class EmailQuerySerializer @Inject()(mailboxIdFactory: MailboxId.Factory) {
     case _ => JsError(s"Expecting a JsString to represent a known operator")
   }
 
-  private implicit val filterConditionReads: Reads[FilterCondition] = Json.reads[FilterCondition]
+  private implicit val filterConditionReads: Reads[FilterCondition] = {
+    case JsObject(underlying) => {
+      val unsupported: collection.Set[String] = underlying.keySet.diff(FilterCondition.SUPPORTED)
+      if (unsupported.nonEmpty) {
+        JsError(s"These '${unsupported.reduce((option, other) => String.format("[%s, %s]", option, other))}' was unsupported filter options")
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/EmailQuerySerializer.scala:88]: maybe unsupported.mkString is more simple than the reduce operation here?

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
```
MemoryMailboxQueryMethodTest.shouldReturnInvalidArgumentsWhenInvalidFilterCondition » IllegalArgument

[b0fa3d2463f014de25008593c17def63761f47e1] [ERROR] shouldReturnInvalidArgumentsWhenInvalidFilterCondition  Time elapsed: 0.248 s  <<< ERROR!
[b0fa3d2463f014de25008593c17def63761f47e1] java.lang.IllegalArgumentException: Can not parse expected value.
[b0fa3d2463f014de25008593c17def63761f47e1] Caused by: com.fasterxml.jackson.core.JsonParseException: 
[b0fa3d2463f014de25008593c17def63761f47e1] Unexpected character ('p' (code 112)): was expecting comma to separate Object entries
[b0fa3d2463f014de25008593c17def63761f47e1]  at [Source: (StringReader); line: 3, column: 35]
```

A simple JSON formatting issue

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**